pymc3/model.py

```diff
@@ -161,49 +162,130 @@ def _get_named_nodes_and_relations(graph, parent, leaf_nodes,
             node_children.update(temp_tree)
     return leaf_nodes, node_parents, node_children
 
+T = TypeVar('T', bound='ContextMeta')
 
-class Context:
+
+class ContextMeta(type):
     """Functionality for objects that put themselves in a context using
     the `with` statement.
     """
-    contexts = threading.local()
-
-    def __enter__(self):
-        type(self).get_contexts().append(self)
-        # self._theano_config is set in Model.__new__
-        if hasattr(self, '_theano_config'):
-            self._old_theano_config = set_theano_conf(self._theano_config)
-        return self
 
-    def __exit__(self, typ, value, traceback):
-        type(self).get_contexts().pop()
-        # self._theano_config is set in Model.__new__
-        if hasattr(self, '_old_theano_config'):
-            set_theano_conf(self._old_theano_config)
-
-    @classmethod
-    def get_contexts(cls):
-        # no race-condition here, cls.contexts is a thread-local object
+    def __new__(cls, name, bases, dct,  **kargs):
+        # this serves only to strip off keyword args, per the warning from
+        # StackExchange:
+        # DO NOT send "**kargs" to "type.__new__".  It won't catch them and
+        # you'll get a "TypeError: type() takes 1 or 3 arguments" exception.   
+        return super().__new__(cls, name, bases, dct)
+
+    # FIXME: is there a more elegant way to automatically add methods to the class that
```

Comment:
Yes, you can add the methods dynamically inside `__new__` by appending them to the `dct`. Check out [this SO question](https://stackoverflow.com/q/32648767/7253166) and [this blog post on how to add them as bound method](https://www.ianlewis.org/en/dynamically-adding-method-classes-or-class-instanc).

Reply:
A couple of questions:

1. I can't decide whether it's better to do this in `__new__` or `__init__`, or how I would make that decision.  Is there some reason it would be better to do this in `__new` than in `__init__`, the way I do?
2. Doing this in `__new__` by modifying the dict seems no nicer, syntactically, than what I do.  What I was looking for was a way to do this that would make my intentions clear to a maintainer reading my code afterwards.  Neither of these alternatives seems very nice, because they read as so different from adding a method in a normal class declaration.  In fact, I think messing with the dictionary might be even less readable than the assignment on lines 201 and 202 below.

Reply:
I'm not sure if it's better to use `__new__` or `__init__` in our case. I think that using `__new__` should be preferred because the newly instantiated class will have all of its methods available when `__init__` is called. This could be useful if someone ever wants to write a new metaclass that inherits from `ContextMeta`, and shouldn't have to worry about when `supe().__init__(...)` gets called. Of course, I think that this is highly unlike to happen in our case, but I think that given that doing things in `__new__` or `__init__` is almost the same, we can just go for `__new__`.

Reply:
OK, I'll make the change, then if the tests pass, I'll merge.  I'm kind of over-extended at work right now, so it might take me a little time.